src/main/java/org/tikv/txn/gc/GCWorker.java

```diff
@@ -0,0 +1,492 @@
+  private ExecutorService gcTaskThreadPool;
+  private ExecutorCompletionService<GCTask> gcTaskService;
+
+  private static final ByteString GC_LAST_RUN_TIME_KEY = ByteString.copyFromUtf8("tikv_gc_last_run_time");
+  private static final ByteString GC_RUN_INTERVAL_KEY = ByteString.copyFromUtf8("tikv_gc_run_interval");
+  private static final ByteString GC_LIFE_TIME_KEY = ByteString.copyFromUtf8("tikv_gc_life_time");
+  private static final ByteString GC_SAFE_POINT_KEY = ByteString.copyFromUtf8("tikv_gc_safe_point");
+  private static final ByteString GC_CONCURRENCY_KEY = ByteString.copyFromUtf8("tikv_gc_concurrency");
+  private static final ByteString GC_LEADER_UUID_KEY = ByteString.copyFromUtf8("tikv_gc_leader_uuid");
+  private static final ByteString GC_LEADER_LEASE_KEY = ByteString.copyFromUtf8("tikv_gc_leader_lease");
+  public static final int gcScanLockLimit = 1024;
+  private final int gcDefaultRunInterval;// = 6 * 1000
+  private final int gcDefaultLifeTime;// = 2 * 1000;
+  private final int gcDefaultConcurrency;
+  private static final int gcMinConcurrency = 1;
+  private static final int gcMaxConcurrency = 128;
+  private final int gcWaitTime;// = 3 * 1000;
+  private final int gcWorkerLease;// = 2 * 1000;
+
+  public GCWorker(TiSession session) {
+    this.pdClient = session.getPDClient();
+    TiConfiguration conf = session.getConf();
+    this.gcDefaultRunInterval = conf.getGCRunInterval();
+    this.gcDefaultLifeTime = conf.getGCLifeTime();
+    this.gcDefaultConcurrency = conf.getGCConcurrency();
+    this.gcWaitTime = conf.getGCWaitTime();
+    this.gcWorkerLease = conf.getGCWorkerLease();
+    this.l_uuid = pdClient.grantLease(gcWorkerLease / 1000);
+    this.uuid = Long.toHexString(l_uuid);
+    logger.info("uuid = " + uuid + " " + l_uuid);
+    this.lastFinish = 0;
+    this.regionStoreClientBuilder = new RegionStoreClientBuilder(session.getConf(), session.getChannelFactory(), new RegionManager(session.getPDClient()));
+  }
+
+  @Override
+  public void close() throws Exception {
+    pdClient.unlock(GC_LEADER_UUID_KEY);
+    pdClient.revoke(l_uuid);
+    if (gcTaskThreadPool != null) {
+      gcTaskThreadPool.shutdown();
+    }
+    scheduler.shutdown();
+    gcWorkerScheduler.shutdown();
+  }
+
+  private TiTimestamp getTimestamp() {
+    return pdClient.getTimestamp(ConcreteBackOffer.newTsoBackOff());
+  }
+
+  private long getNow() {
+    return getTimestamp().getPhysical();
+  }
+
+  // Note: Should not call start() more than once
+  public void start() {
+    gcWorkerScheduler.schedule(this::run, 0, TimeUnit.MILLISECONDS);
+  }
+
+  private void run() {
+    logger.info(String.format("[gc worker] %s starts", uuid));
+    scheduler.scheduleAtFixedRate(() -> {
+      try {
+        logger.debug("[gc worker] tick");
+        pdClient.keepLeaseAlive(l_uuid);
+        tick();
+      } catch (Exception e) {
+        logger.warn("[gc worker] gc fails to proceed", e);
+      }
+    }, 0, 1, TimeUnit.SECONDS);
```

Comment:
Here, if a tick doesn't finish for a long time, will the next tick run asynchronously? If it's not so, then the `runGCJob` at L164 should be asynchronous. Sorry I didn't notice this question in my previous review.

Reply:
JavaDoc
> If any execution of this task takes longer than its period, then subsequent executions may start late, but will not concurrently execute.

Reply:
Oh, then the `runGCJob` at L164 should be a async task. When the db is very large it's possible to take very long time to finish the whole gc, and the tick shouldn't stop.